docs/arch/Proposals/adr-2504-bulkupdate.md

```diff
@@ -0,0 +1,172 @@
+
+### Decision
+- Server processes only one bulk-update job at a time. If a new job is submitted while an existing one is in progress, the request will be rejected with a valid reason provided to the customer. This prevents conflicts during resource updates that could corrupt the data.
+- The bulk-update endpoint accepts an `isParallel` flag, which defaults to `true`. When isParallel is set to true and there are no search parameters in the request, the BulkUpdate orchestrator creates processing jobs on partitions based on `resource type and surrogate ID ranges`, similar to the export job. These jobs run in `parallel` to maximize throughput.
+- If isParallel is set to true and search parameters are included in the request, the BulkUpdate orchestrator reads all pages of data, including the included results, and creates processing jobs at each `continuation token level`. These jobs also run in `parallel` to maximize throughput.
+- When isParallel is set to false, the BulkUpdate orchestrator enqueues a single processing job that reads each page of data and performs updates sequentially. This approach results in relatively lower throughput.
+- A single transaction will update `1,000` resources using the Merge stored procedure, similar to other background jobs.
+- Updates on certain resource types could render the user's database invalid. To prevent this, specific resource types like `SearchParameter and ValueSet` will be excluded from bulk-update operations at both the system and resource type levels. (Previously, issues with ValueSet updates were encountered during bundle operations.)
+- Every patch `path` **must start with the `ResourceType` root** (e.g., `Patient.meta.tag`) to disambiguate meta‑level vs element updates. 
+- To patch the `Common properties` users should use `Resource`. (e.g., `"name": "path", "valueString": "Resource"`)
+- The bulk-update job must run to completion, with retry logic in place to handle transient, connectivity, or availability failures.
+- Transient, connectivity, or availability failures during job processing should not result in duplicate entries being added to a resource. To prevent this, updates should be processed in chunks wrapped within a `transaction`. 
+- If the user cancels the job, any updates that have already been committed `cannot be rolled back`.
+- A resource will only be updated, along with its metadata (e.g., version and last update time), when an actual change occurs.
+- If the update request results in an invalid resource, the update will fail for all resources of the same type. The count of resources that failed to patch will be reported to the user. User can run the regular FHIR Patch on an individual resource to get the detailed error message.
+- Attempting to `replace` an element that does not exist will fail in the FHIR Patch call. The same behavior applies to the bulk-update operation.
```

Comment:
I think it is important to mention that the job won't fail, only the specific update to that resource will fail. Since this is the first job type with the concept of soft failure I think it is worth calling out.

Reply:
I agree